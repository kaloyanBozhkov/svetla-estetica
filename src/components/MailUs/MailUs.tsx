import React, { useReducer } from 'react'

// import common types
import { MailForm } from '~/common/types'

// import helpers
import sendMail from '~/services/sendMail'
import { validatePhoneNumber, validateEmail } from '~/helpers/validators'
import useTimerToggle from '~/hooks/useTimerToggle'

// import atoms
import Button from '../UI/Button/Button'
import Input from '../UI/Input/Input'

// import styles
import styles from './mailUs.module.scss'

type action = {
    fieldName: string
    fieldValue: string
}

const valueUpdater = (fieldName: string, setInputs: (action: action) => void) => ({
        target: { value: fieldValue },
    }: React.ChangeEvent<HTMLInputElement | HTMLTextAreaElement>) =>
        setInputs({ fieldName, fieldValue }),
    fields: Array<{ label: string; property: keyof MailForm; isValid?: (s: string) => boolean }> = [
        {
            label: 'Nome',
            property: 'name',
        },
        {
            label: 'Cognome',
            property: 'surname',
        },
        {
            label: 'Email',
            property: 'email',
            isValid: validateEmail,
        },
        {
            label: 'Telefono',
            property: 'phone',
            isValid: validatePhoneNumber,
        },
    ]

const MailUs = (): JSX.Element => {
    const [inputs, setInputs] = useReducer(
            (acc: MailForm, action: action) => ({
                ...acc,
                [action.fieldName]: action.fieldValue,
            }),
            {
                name: '',
                surname: '',
                email: '',
                phone: '',
                message: '',
            }
        ),
        [toggle, setToggled] = useTimerToggle({ untoggleAfter: 1 }),
        cantSubmit = !!Object.values(inputs).filter((i) => !i).length

    return (
        <div className={styles.mailUs}>
            <div className={styles.container}>
                {fields.map(({ label, property, isValid }, index) => {
                    const isValidField =
                        !!inputs[property] && (isValid ? isValid(inputs[property]) : true)

                    return (
                        <div
                            key={index}
                            className={styles.inputWrapper}
                            data-is-valid={isValidField}
                            data-is-toggled={toggle}
                        >
                            <Input
                                label={label}
                                value={inputs[property]}
                                onChange={valueUpdater(property, setInputs)}
                            />
                        </div>
                    )
                })}
            </div>
            <div
                className={styles.inputWrapper}
                data-is-valid={!!inputs.message}
                data-is-toggled={toggle}
            >
                <Input
                    label="Messaggio"
                    type="text"
                    value={inputs.message}
                    onChange={valueUpdater('message', setInputs)}
                />
            </div>
            <div className={styles.btnWrapper} data-is-disabled={cantSubmit}>
                <Button
                    modifier="solid"
                    label="Invia"
                    onClick={() => {
                        if (cantSubmit) {
                            // enable showing toggled
                            return !toggle && setToggled()
                        }

                        sendMail(inputs)
                    }}
                />
            </div>
        </div>
    )
}

export default MailUs
