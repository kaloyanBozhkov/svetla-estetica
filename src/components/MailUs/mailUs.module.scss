@import "scss/mixins.scss";
@import "scss/variables.scss";

.mailUs {
    @include sidePaddings;
    padding-top: 5rem;
    padding-bottom: 5rem;

    .container {
        display: grid;
        grid-template-columns: 1fr 1fr;
        grid-template-rows: 5rem;
        column-gap: 23rem;
        row-gap: 5rem;
        margin-bottom: 5rem;

        .inputWrapper {
            height: 4.5rem;
        }
    }

    > .inputWrapper {
        height: 18rem;
        margin-bottom: 5rem;
    }

    > .btnWrapper {
        height: 5rem;
        width: 20rem;

        button {
            font-weight: bold;
            font-size: 1.8rem;
        }

        &[data-is-disabled="false"] {
            button {
                color: $green;
                border-color: $green;
                &:hover {
                    background-color: $green;
                    color: white;
                }
            }
        }
    }

    .inputWrapper[data-is-valid="true"][data-is-toggled="false"] {
        input,
        textarea {
            border-color: $green;

            ~ label {
                color: $green;
            }
        }
    }

    .inputWrapper[data-is-valid="false"][data-is-toggled="true"] {
        @keyframes shake {
            0% {
                transform: rotate(0deg);
            }
            25% {
                transform: rotate(2deg);
            }
            75% {
                transform: rotate(-2deg);
            }
            100% {
                transform: rotate(0deg);
            }
        }
        @include animate(shake, 330ms);
        animation-iteration-count: infinite;

        input,
        textarea {
            @keyframes borderColor {
                50% {
                    border-color: red;
                }
            }
            @include animate(borderColor, 1s);
        }
    }
}
