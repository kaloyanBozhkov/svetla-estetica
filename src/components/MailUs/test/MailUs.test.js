import React from 'react'
import { shallow } from 'enzyme'
import { act } from 'react-dom/test-utils'

// import components
import MailUs from '../MailUs'

// mock the helper
jest.mock('services/sendMail')

describe('testing MailUs', () => {
    it('should render component', () => {
        expect(shallow(<MailUs />)).toMatchSnapshot()
    })

    it('should handle setting input values', () => {
        const wrapper = shallow(<MailUs />)
        let inputs = wrapper.find('Input'),
            onNameInputChange = inputs.at(0).prop('onChange')

        expect(inputs.at(0).prop('value')).toBe('')

        act(() => {
            onNameInputChange({ target: { value: 'hello' } })
        })

        wrapper.update()

        inputs = wrapper.find('Input').at(0)

        expect(inputs.prop('value')).toBe('hello')
    })

    it('should disable btn and not run sendMail if not all inputs are set', () => {
        const wrapper = shallow(<MailUs />),
            btn = wrapper.find('Button')

        expect(btn.parent().prop('data-is-disabled')).toBe(true)

        btn.prop('onClick')()

        expect(sendMail).toHaveBeenCalledTimes(0)
    })

    it('should disable btn and not run sendMail if not all inputs are set', () => {
        const wrapper = shallow(<MailUs />),
            inputs = wrapper.find('Input')
        let btn = wrapper.find('Button')

        expect(sendMail).toHaveBeenCalledTimes(0)

        expect(btn.parent().prop('data-is-disabled')).toBe(true)

        act(() => {
            inputs.forEach((input) => {
                input.prop('onChange')({ target: { value: 'hello' } })
            })
        })

        wrapper.update()

        btn = wrapper.find('Button')

        expect(btn.parent().prop('data-is-disabled')).toBe(false)

        btn.prop('onClick')()

        expect(sendMail).toHaveBeenCalledTimes(1)
    })
})
