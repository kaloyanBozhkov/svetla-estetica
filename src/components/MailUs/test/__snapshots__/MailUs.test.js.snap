// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing MailUs should render component 1`] = `
<div
  className="mailUs"
>
  <div
    className="container"
  >
    <div
      className="inputWrapper"
      data-is-toggled={false}
      data-is-valid={false}
      key="0"
    >
      <Input
        label="Nome"
        onChange={[Function]}
        value=""
      />
    </div>
    <div
      className="inputWrapper"
      data-is-toggled={false}
      data-is-valid={false}
      key="1"
    >
      <Input
        label="Cognome"
        onChange={[Function]}
        value=""
      />
    </div>
    <div
      className="inputWrapper"
      data-is-toggled={false}
      data-is-valid={false}
      key="2"
    >
      <Input
        label="Email"
        onChange={[Function]}
        value=""
      />
    </div>
    <div
      className="inputWrapper"
      data-is-toggled={false}
      data-is-valid={false}
      key="3"
    >
      <Input
        label="Telefono"
        onChange={[Function]}
        value=""
      />
    </div>
  </div>
  <div
    className="inputWrapper"
    data-is-toggled={false}
    data-is-valid={false}
  >
    <Input
      label="Messaggio"
      onChange={[Function]}
      type="text"
      value=""
    />
  </div>
  <div
    className="btnWrapper"
    data-is-disabled={true}
  >
    <Button
      label="Invia"
      modifier="solid"
      onClick={[Function]}
    />
  </div>
</div>
`;
